enhancements/cluster-api/crd-compatibility-checker.md

```diff
@@ -0,0 +1,469 @@
+
+## Proposal
+
+The proposal introduces a new admission webhook for CRDs.
+We introduce a new cluster-scoped CRD called `CRDCompatibilityRequirement` that allows actors to express compatibility requirements for target CRDs.
+The admission webhook ensures that subsequent changes to referenced CRDs are compatible with all requirements specified in a `CRDCompatibilityRequirement`.
+
+### Workflow description
+
+The following are 2 related example uses of the proposed API and controller.
+
+#### CCAPIO ensures cluster stability while HCP manages CAPI CRDs
+
+In this example, Hosted Control Planes (HCP) takes responsibility for the management of the cluster-api (CAPI) `Cluster` CRD from cluster-capi-operator (CCAPIO).
+
+CCAPIO deploys CAPI version X and uses it to manage cluster resources.
+
+HCP deploys CAPI version Y, where Y>X. HCP uses it to deploy external managed clusters.
+
+As `Cluster` can only have a single CRD in the management cluster, they must both use the same CRD.
+
+* HCP will assert ownership of the `Cluster` CRD
+* CCAPIO will assert its requirements for the `Cluster` CRD, but will no longer manage the CRD.
+* HCP will upgrade the `Cluster` CRD
+
+CRD Compatibility Checker ensures that the version HCP upgrades to does not break CCAPIO, and therefore the management cluster.
+
+During normal operation, CCAPIO applies sets of resources to the cluster which are defined in 'transport' config maps.
+CCAPIO will be enhanced such that it will automatically create a CRDCompatibilityRequirement for any CRD it loads.
+In the case of `Cluster` this would be:
+
+```yaml
+apiVersion: operator.openshift.io/v1alpha1
+kind: CRDCompatibilityRequirement
+metadata:
+  name: cluster-api-cluster-ccapio
+spec:
+  crdRef: clusters.cluster.x-k8s.io
+  creatorDescription: "OpenShift Cluster CAPI Operator"
+  compatibilityCRD: |
+    ...
+    <complete YAML document of Cluster CRD from transport config map>
+    ...
+  crdAdmitAction: Enforce
+  objectSchemaValidation:
+    namespaceSelector:
+      matchLabels:
+        "kubernetes.io/metadata.name": "openshift-cluster-api"
+    action: Enforce
+```
+
+CCAPIO marks itself Degraded if any CRDCompatibilityRequirement it creates during this process becomes not Progressing, and:
+* not Admitted - transport config map is invalid
+* not Compatible - admitted CRD is incompatible with CCAPIO requirements
+
+HCP uses an as-yet-undefined mechanism to inform CCAPIO of the set of CRDs which are no longer managed by CCAPIO.
+For CRDs in this list, CCAPIO creates a CRDCompatibilityRequirement but does not load or update the CRD.
+This is the only operational change required by HCP.
+
+HCP proceeds normally and upgrades the `Cluster` CRD.
+Assuming that the CRD from version Y is compatible with the one for version X required by CCAPIO, the CRD is admitted and everything works as normal.
+
+#### CCAPIO prevents cluster upgrade if it would conflict with an existing CRD
```

Comment:
I had expected this to be in the other EP, but, I guess it doesn't matter if it's duplicated, adds context on why certain decisions were made around the ability to warn. Any concern about the duplication?